### ISSUE TITLE
Test::BDD::Cucumber tests broken since 0.61

### ISSUE THREAD

**Issue by jwrightecs** [pherkin/test-bdd-cucumber-perl#155]
```
$ make test
PERL_DL_NONLAZY=1 "/usr/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/210_background_sections.t ....................... ok    
t/220_tag_parsing.t ............................... ok   
t/230_tag_matching.t .............................. ok   
t/240_localized_features.t ........................ ok   
t/250_i18n_json.t ................................. ok   
t/260_match_matcher.t ............................. ok   
t/270_scenario_outlines.t ......................... ok    
t/310_auto_corpus.t ............................... ok   
t/400_app_pherkin_harnesses.t ..................... ok   
t/500_error_formatter.t ........................... ok   
t/600_harness_json_output.t ....................... ok    
t/700_tag_processing.t ............................ ok   
t/710_configuration_profiles.t .................... ok    
t/720_extension_tests.t ........................... ok   
t/800_regressions_010_too_few_features.t .......... ok   
t/800_regressions_020_term_color_skipped_steps.t .. ok   
t/800_regressions_030_pherkin_external_libs.t ..... ok   
t/900_run_cucumber_tests.t ........................ 10/? 
# Failed test 'In Then the scenario is pending'
# at /home/perl/.cpanm/work/1568644874.11936/Test-BDD-Cucumber-0.64/blib/lib/Test/BDD/Cucumber/Executor.pm line 643.
#   in step at t/cucumber_core_features/core.feature line 48.
#     ok - Starting to execute step: the scenario is pending
#     not ok 1 - Scenario status is pending
#     #   Failed test 'Scenario status is pending'
#     #   at /home/jwright/perl5/lib/perl5/Test/Builder.pm line 334.
#     #          got: 'failing'
#     #     expected: 'pending'
#     #     ok - Starting to execute step: I add 4 and 5
#     #     not ok 1 - Todo Step
#     #     #   Failed test 'Todo Step'
#     #     #   at /home/perl/perl5/lib/perl5/Test/Builder.pm line 334.
#     #     1..2
#     #     # Looks like you failed 1 test of 2.
#     # 
#     # 1..0 # SKIP Short-circuited from previous tests
#     # 
#     1..2
#     # Looks like you failed 1 test of 2.
# Looks like you failed 1 test of 271.
t/900_run_cucumber_tests.t ........................ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/271 subtests 
	(less 1 skipped subtest: 269 okay)
t/910_run_tests_pherkin.t ......................... ok   
t/author-pod-syntax.t ............................. skipped: these tests are for testing by the author

Test Summary Report
-------------------
t/900_run_cucumber_tests.t                      (Wstat: 256 Tests: 271 Failed: 1)
  Failed test:  228
  Non-zero exit status: 1
Files=20, Tests=427,  4 wallclock secs ( 0.10 usr  0.03 sys +  3.78 cusr  0.30 csys =  4.21 CPU)
Result: FAIL
Failed 1/20 test programs. 1/427 subtests failed.
make: *** [Makefile:1003: test_dynamic] Error 255

```
Tested on 5.10.1, 5.30.0 and 5.28.1

**Comment by ehuelsmann** [comment]
Hi,

Yes, this is a problem I noticed earlier today with the CPAN testers too. The problem seems related to the version of Test-Simple that's installed on the tester: versions 1.302162 and higher seem to fail; version 1.302086 and up (mostly being tested: 1.302133) seem to work.

The hint that the output above provides is that this is about $TODO handling. I'm not sure this is a problem in Test::BDD::Cucumber or an issue with a change in Test-Simple.

@jwrightecs which version of Test-Simple (Test::More) did you have installed?

Regards,

**Comment by andk** [comment]
Hi, you can see statistics of the cpan testers runs on http://analysis.cpantesters.org/solved?distv=Test-BDD-Cucumber-0.64

They show it is always the test t/900_run_cucumber_tests.t that fails. All other statistics have weak statistical scores and probably do not help finding the culprit.

What would interest me: more output from the failing test. From what I see it seems to fail the test that 4 plus 5 is 9. How does it draw that conclusion? Maybe a wrong comparison operator? Maybe more intermediate tests would reveal more.

**Comment by ehuelsmann** [comment]
hi @andk, @jwrightecs,

Thanks for looking into this. The output is a bit misleading, because the .. [The full test being run](https://github.com/pherkin/test-bdd-cucumber-perl/blob/master/t/cucumber_core_features/core.feature#L39-L49) is:

```gherkin
  Scenario: Pending step means the scenario is pending
    Given a scenario with:
      """
      When I add 4 and 5
      Then the result is 9
      """
    And the step "I add 4 and 5" has a pending mapping
x   And the step "the result is 9" has a passing mapping
    When Cucumber executes the scenario
*   Then the scenario is pending
    And the step "the result is 9" is skipped
```

The failing step is the one marked with the `*`. The step marked with the `x` runs this code https://github.com/pherkin/test-bdd-cucumber-perl/blob/master/t/cucumber_core_features/step_definitions/core_steps.pl#L44-L48 which attaches [a Test::More function to that step which should cause Test::More to interpret that step as TODO](https://github.com/pherkin/test-bdd-cucumber-perl/blob/master/t/cucumber_core_features/step_definitions/core_steps.pl#L18-L20):

```perl
my $step_mappings = {
    passing => sub { pass("Step passed") },
    failing => sub { fail("Step passed") },
    pending => sub {
      TODO: { local $TODO = 'Todo Step'; ok( 0, "Todo Step" ) }
    }
};
```
The `pending` function.

These lines didn't change with any of the recent releases. Locally, I don't have any failures at all. I'm wondering if http://analysis.cpantesters.org/solved?distv=Test-BDD-Cucumber-0.64#mod:Test::More indicates that there's a correlation with recent Test-Simple releases and that this might be a bug due to weird interaction of Test::BDD::Cucumber with Test2?

**Comment by ehuelsmann** [comment]
What I mean to say is that Test::BDD::Cucumber converts a `TODO: { local $TODO='todo'; ok(0) }` to `pending`. Apparently, in some versions of Test-Simple (Test2), the TODO block doesn't get detected correctly (or doesn't get included in output correctly...).

**Comment by andk** [comment]
> wondering if http://analysis.cpantesters.org/solved?distv=Test-BDD-Cucumber-0.64#mod:Test::More indicates that there's a correlation with recent Test-Simple releases

There is a correlation but the low R^2 suggests it is unlikely to be interesting.

Since the fails started with 0.61 (the subject line of this issue is
not my observation): maybe you can try to revert the changes between
0.60 and 0.61 and then slowly find the prove what went wrong with the
switch to Test2?

**Comment by ehuelsmann** [comment]
You're right with your observation. I've changed the title.

**Comment by ehuelsmann** [comment]
As to going back to 0.60 and stepping through the commits until I find the problem -- I'd love to, but my system is among the ones which have fully succeeding tests and since I don't know what the cause of the issue is on the cpantesters where it fails (the tests don't fail on all cpantesters), I have no idea what I need to do to reproduce...

**Comment by ehuelsmann** [comment]
To be more specific: my local environment is perl 5.10.1 and 5.24.4; both with Test2::API / Test::More 1.302167. Upgrading 5.24.4 to 1.302168 didn't make any errors show up.

Also, the Travis CI tests (https://travis-ci.org/pherkin/test-bdd-cucumber-perl/builds/587363277) don't have any issues. I'm thinking there's some indirect dependency which influences these results. I'll see if I can run some tests through cpantesters which include indirect dependents as direct ones in order to have them included in the variance analysis.

**Comment by jwrightecs** [comment]
Unfortunately the commit that seems to break it is https://github.com/pherkin/test-bdd-cucumber-perl/commit/aece3e31fd9d8fb3608e1343ce3269b610c7e735 so it might need to be broken into smaller pieces. 

Running cpan Test::BDD::Cucumber on the perl:latest image on dockerhub 

```
$ docker run -it perl cpan Test::BDD::Cucumber
... lots of cpan installs ..
Running make test for EHUELS/Test-BDD-Cucumber-0.64.tar.gz
PERL_DL_NONLAZY=1 "/usr/local/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/210_background_sections.t ....................... ok    
t/220_tag_parsing.t ............................... ok   
t/230_tag_matching.t .............................. ok   
t/240_localized_features.t ........................ ok   
t/250_i18n_json.t ................................. ok   
t/260_match_matcher.t ............................. ok   
t/270_scenario_outlines.t ......................... ok    
t/310_auto_corpus.t ............................... ok   
t/400_app_pherkin_harnesses.t ..................... ok   
t/500_error_formatter.t ........................... ok   
t/600_harness_json_output.t ....................... ok    
t/700_tag_processing.t ............................ ok   
t/710_configuration_profiles.t .................... ok    
t/720_extension_tests.t ........................... ok   
t/800_regressions_010_too_few_features.t .......... ok   
t/800_regressions_020_term_color_skipped_steps.t .. ok   
t/800_regressions_030_pherkin_external_libs.t ..... ok   
t/900_run_cucumber_tests.t ........................ 1/? 
# Failed test 'In Then the scenario is pending'
# at /root/.cpan/build/Test-BDD-Cucumber-0.64-0/blib/lib/Test/BDD/Cucumber/Executor.pm line 643.
#   in step at t/cucumber_core_features/core.feature line 48.
#     ok - Starting to execute step: the scenario is pending
#     not ok 1 - Scenario status is pending
#     #   Failed test 'Scenario status is pending'
#     #   at /usr/local/lib/perl5/5.30.0/Test/Builder.pm line 333.
#     #          got: 'failing'
#     #     expected: 'pending'
#     #     ok - Starting to execute step: I add 4 and 5
#     #     not ok 1 - Todo Step
#     #     #   Failed test 'Todo Step'
#     #     #   at /usr/local/lib/perl5/5.30.0/Test/Builder.pm line 333.
#     #     1..2
#     #     # Looks like you failed 1 test of 2.
#     # 
#     # 1..0 # SKIP Short-circuited from previous tests
#     # 
#     1..2
#     # Looks like you failed 1 test of 2.
t/900_run_cucumber_tests.t ........................ 187/? # Looks like you failed 1 test of 271.
t/900_run_cucumber_tests.t ........................ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/271 subtests 
        (less 1 skipped subtest: 269 okay)
t/910_run_tests_pherkin.t ......................... ok   
t/author-pod-syntax.t ............................. skipped: these tests are for testing by the author

Test Summary Report
-------------------
t/900_run_cucumber_tests.t                      (Wstat: 256 Tests: 271 Failed: 1)
  Failed test:  185
  Non-zero exit status: 1
Files=20, Tests=427,  5 wallclock secs ( 0.11 usr  0.02 sys +  4.22 cusr  0.41 csys =  4.76 CPU)
Result: FAIL
Failed 1/20 test programs. 1/427 subtests failed.
make: *** [Makefile:937: test_dynamic] Error 255
  EHUELS/Test-BDD-Cucumber-0.64.tar.gz
  /usr/bin/make test -- NOT OK
//hint// to see the cpan-testers results for installing this module, try:
  reports EHUELS/Test-BDD-Cucumber-0.64.tar.gz
$

```

**Comment by ehuelsmann** [comment]
Thanks! Checking if that fails locally too. Building now.

**Comment by ehuelsmann** [comment]
GAH! it's not failing locally:

```plain
  EHUELS/Test-BDD-Cucumber-0.64.tar.gz
  Has already been unwrapped into directory /root/.cpan/build/Test-BDD-Cucumber-0.64-0
  EHUELS/Test-BDD-Cucumber-0.64.tar.gz
  Has already been prepared
Running make for E/EH/EHUELS/Test-BDD-Cucumber-0.64.tar.gz
cp lib/Test/BDD/Cucumber/Harness/TAP.pm blib/lib/Test/BDD/Cucumber/Harness/TAP.pm
cp lib/Test/BDD/Cucumber/StepContext.pm blib/lib/Test/BDD/Cucumber/StepContext.pm
cp lib/Test/BDD/Cucumber/I18n.pm blib/lib/Test/BDD/Cucumber/I18n.pm
cp lib/App/pherkin.pm blib/lib/App/pherkin.pm
cp lib/Test/BDD/Cucumber/Errors.pm blib/lib/Test/BDD/Cucumber/Errors.pm
cp lib/Test/BDD/Cucumber/Harness/JSON.pm blib/lib/Test/BDD/Cucumber/Harness/JSON.pm
cp lib/Test/BDD/Cucumber/I18N/Data.pm blib/lib/Test/BDD/Cucumber/I18N/Data.pm
cp lib/Test/BDD/Cucumber/Model/Result.pm blib/lib/Test/BDD/Cucumber/Model/Result.pm
cp lib/Test/BDD/Cucumber/Harness/TestBuilder.pm blib/lib/Test/BDD/Cucumber/Harness/TestBuilder.pm
cp lib/Test/BDD/Cucumber/Harness.pm blib/lib/Test/BDD/Cucumber/Harness.pm
cp lib/Test/BDD/Cucumber/Model/TagSpec.pm blib/lib/Test/BDD/Cucumber/Model/TagSpec.pm
cp lib/Test/BDD/Cucumber/Parser.pm blib/lib/Test/BDD/Cucumber/Parser.pm
cp lib/Test/BDD/Cucumber/Manual/Steps.pod blib/lib/Test/BDD/Cucumber/Manual/Steps.pod
cp lib/Test/BDD/Cucumber/Model/Document.pm blib/lib/Test/BDD/Cucumber/Model/Document.pm
cp lib/Test/BDD/Cucumber/Model/Scenario.pm blib/lib/Test/BDD/Cucumber/Model/Scenario.pm
cp lib/Test/BDD/Cucumber/StepFile.pm blib/lib/Test/BDD/Cucumber/StepFile.pm
cp lib/Test/BDD/Cucumber/Manual/Integration.pod blib/lib/Test/BDD/Cucumber/Manual/Integration.pod
cp lib/Test/BDD/Cucumber/Loader.pm blib/lib/Test/BDD/Cucumber/Loader.pm
cp lib/Test/BDD/Cucumber/Util.pm blib/lib/Test/BDD/Cucumber/Util.pm
cp lib/Test/BDD/Cucumber/Harness/TermColor.pm blib/lib/Test/BDD/Cucumber/Harness/TermColor.pm
cp lib/Test/BDD/Cucumber/Executor.pm blib/lib/Test/BDD/Cucumber/Executor.pm
cp lib/Test/BDD/Cucumber/Harness/Data.pm blib/lib/Test/BDD/Cucumber/Harness/Data.pm
cp lib/Test/BDD/Cucumber/Model/Feature.pm blib/lib/Test/BDD/Cucumber/Model/Feature.pm
cp lib/Test/BDD/Cucumber/Model/Step.pm blib/lib/Test/BDD/Cucumber/Model/Step.pm
cp lib/Test/BDD/Cucumber/Manual/Tutorial.pod blib/lib/Test/BDD/Cucumber/Manual/Tutorial.pod
cp lib/Test/BDD/Cucumber/Model/Line.pm blib/lib/Test/BDD/Cucumber/Model/Line.pm
cp lib/TAP/Parser/Iterator/PherkinStream.pm blib/lib/TAP/Parser/Iterator/PherkinStream.pm
cp lib/TAP/Parser/SourceHandler/Feature.pm blib/lib/TAP/Parser/SourceHandler/Feature.pm
cp lib/Test/BDD/Cucumber/Extension.pm blib/lib/Test/BDD/Cucumber/Extension.pm
cp lib/Test/BDD/Cucumber/Manual/Architecture.pod blib/lib/Test/BDD/Cucumber/Manual/Architecture.pod
cp lib/Test/BDD/Cucumber.pm blib/lib/Test/BDD/Cucumber.pm
cp bin/pherkin blib/script/pherkin
"/usr/local/bin/perl" -MExtUtils::MY -e 'MY->fixin(shift)' -- blib/script/pherkin
  EHUELS/Test-BDD-Cucumber-0.64.tar.gz
  /usr/bin/make -- OK
Running make test for EHUELS/Test-BDD-Cucumber-0.64.tar.gz
PERL_DL_NONLAZY=1 "/usr/local/bin/perl" "-MExtUtils::Command::MM" "-MTest::Harness" "-e" "undef *Test::Harness::Switches; test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/210_background_sections.t ....................... ok    
t/220_tag_parsing.t ............................... ok   
t/230_tag_matching.t .............................. ok   
t/240_localized_features.t ........................ ok   
t/250_i18n_json.t ................................. ok   
t/260_match_matcher.t ............................. ok   
t/270_scenario_outlines.t ......................... ok    
t/310_auto_corpus.t ............................... ok   
t/400_app_pherkin_harnesses.t ..................... ok   
t/500_error_formatter.t ........................... ok   
t/600_harness_json_output.t ....................... ok    
t/700_tag_processing.t ............................ ok   
t/710_configuration_profiles.t .................... ok    
t/720_extension_tests.t ........................... ok   
t/800_regressions_010_too_few_features.t .......... ok   
t/800_regressions_020_term_color_skipped_steps.t .. ok   
t/800_regressions_030_pherkin_external_libs.t ..... ok   
t/900_run_cucumber_tests.t ........................ ok     
t/910_run_tests_pherkin.t ......................... ok   
t/author-pod-syntax.t ............................. skipped: these tests are for testing by the author
All tests successful.
Files=20, Tests=427,  4 wallclock secs ( 0.11 usr  0.01 sys +  3.36 cusr  0.32 csys =  3.80 CPU)
Result: PASS
  EHUELS/Test-BDD-Cucumber-0.64.tar.gz
  /usr/bin/make test -- OK
```
The test run was initiated using:

```bash
$ docker pull perl:latest
$ docker run --rm -it perl cpan Test::BDD::Cucumber
```

**Comment by andk** [comment]
I just ran bisect on the smokers that can produce fails. I did **not** get `aece3e3` as the culprit but a later one:

```
commit abecf194364b229c817f347a623b972271d329b4
Author: Erik Huelsmann <ehuels@gmail.com>
Date:   Sat Sep 7 12:57:44 2019 +0200

    Add support for step functions using Test2 (with Test::More compat)
    
    Incidentally, this commit also cleans up the hacks which allow to
    grab and redirect Test::Builder's event stream: Test2 has a built-in
    API to grab the test event stream (ok-s and fail-s) which is a perfect
    fit for what we try to do: run an isolated set of tests and present the
    result of those with a specific status to an outer scope test handler.
```

**Comment by ehuelsmann** [comment]
Thanks for the bisect! I was picked by `aece3e3`. Your finding makes much more sense. 

I uploaded a trial dist with a direct dependency on scalar::util, hoping that it provides a differentiating factor (it's one of the very few deps of test-simple). But it turns out not to be the case. I was grasping at straws, obviously.

I'm now wondering if this is a timing issue. But if it is, then how come you can reprice solidly?

**Comment by ehuelsmann** [comment]
The key part is

```plain
#     #          got: 'failing'
#     #     expected: 'pending'
#     #     ok - Starting to execute step: I add 4 and 5
#     #     not ok 1 - Todo Step
```

Especially the last two lines. On non-failing builds, the expected output "pending" is generated, which would have serialized to

```plain
#     #     ok - Starting to execute step: I add 4 and 5
#     #     not ok 1 - Todo Step # TODO Todo Step
```
The fact that the TAP stream doesn't contain the TODO marker makes me believe that in some cases Test-Simple doesn't correctly detect the `$TODO` variable being set. Which is why I was focussing on Test2::API versions as a possible failure condition. But as you suggest, the low R-squared provides little evidence for a correlation.

I'm thinking it would be nice to create a minimal reproduction recipe and submit that to Chad.

**Comment by andk** [comment]
I can reproduce both the passes and the fails, so timing influence is pretty unlikely. From all the facts we have seen so far, it must be some action at a distance that only your test triggers. We cannot say it is a bug in your code, we simply only know that your code triggers something under some architectures. And what we know from the test reports seems not to contain the pointer to the origin.

What I can say about my smokers that exhibit the fails: Ubuntu bionic (uptodate), Debian buster (uptodate) and Debian stretch (not fully uptodate).

The pages about correlations with gccversion at analysis.cpantesters.org give vague hints about which other distros have many fails.

**Comment by ehuelsmann** [comment]
Agreed. Incidentally, I just found a set of changes which makes it fail locally too. I'll upload a trial version to see if that makes it consistent across builders and to set a common baseline to work on to see where the actual issue exists.

**Comment by ehuelsmann** [comment]
Based on the failure case that I now have, I've traced this to an unsupported use-case in Test::Builder. I'm not sure why it failed on some cpantesters and not others, but when I introduce explicit packages in the step files, this is what happens:

```plain
t/900_run_cucumber_tests.t tells `Test::BDD::Cucumber::Loader` to load `t/cucumber_core_features/`
  which causes it to `do` contained files `step_definitions/core_steps.pl` and `tags_steps.pl`
  which each `use Test::More;`
  which - in its import - calls `exported_to` twice; once with `core_steps` and once with `tags_steps` as the package being imported into
```

The second invocation overwrites the `Exported_to` value of the first invocation in `Test::Builder`.

Then the tests get executed. The first test (core.feature) uses the functions defined in the `core_steps` package. None of the functions uses the `$TODO` variable, [except one](https://github.com/pherkin/test-bdd-cucumber-perl/blob/master/t/cucumber_core_features/step_definitions/core_steps.pl#L18-L20). When that function is executed, `Test::Builder` looks at the `$TODO` variable in two packages: `$cpkg` (the caller's package) and `$epkg` (the package indicated by the `Export_To` value in the builder, into which the builder was exported).

Note that the package which invokes the `ok` function in the lines indicated above is the `core_steps` package. The values of `$epkg` and `$cpkg` are respectively **tags_steps** and one of **Test::Builder** (or, if not that, I'm also seeing `Test::More` and `Test::BDD::Cucumber::Executor`, but I think those are related to other invocations).

Conclusion: Test::Builder isn't looking in the correct package to identify the value of the `$TODO` variable.

@andk, @jwrightecs Thanks for helping me identify the problem! I'm sure I'm challenging Test::Builder and Test-Simple here. Based on this finding, I should be able to create a minimal reproduction recipe for Test-Simple and @exodist .

**Comment by ehuelsmann** [comment]
With the analysis above, the following ought to be a minimum reproduction case (but isn't), so there must be another factor at play.

```perl

package p1;

use strict;
use warnings;

use Test::More;

sub t1 {
  local $TODO = 'not implemented yet';

  ok(0, q{yet to implement});
}


package p2;

use strict;
use warnings;

use Test::More;

sub t2 {
   ok(1, 'implemented');

}


package st;

use strict;
use warnings;

require Test::More;

sub s {
   Test::More::subtest 's', sub {
      p1::t1();
   };
}


package t;

use Test::More;

st::s();
p2::t2();

done_testing;
```

**Comment by ehuelsmann** [comment]
I found a minimal reproduction recipe and work-around. I've implemented the workaround in 0.660002-TRIAL and 0.660003-TRIAL (the latter updates the dependencies correctly).

I'll file a ticket with Test-Simple regarding Test2. The reproduction recipe depends solely on Test2 libraries. Once the work-around has been released *and* the issue filed with Test2, I'll close this issue.

**Comment by ehuelsmann** [comment]
I found a minimal reproduction recipe and work-around. I've implemented the workaround in 0.660002-TRIAL and 0.660003-TRIAL (the latter updates the dependencies correctly).

I'll file a ticket with Test-Simple regarding Test2. The reproduction recipe depends solely on Test2 libraries. Once the work-around has been released *and* the issue filed with Test2, I'll close this issue.

**Comment by ehuelsmann** [comment]
0.660003-TRIAL seems to work around it nicely.
